🎨 modernize for 2025 (chores & slight refactoring)

pboling opened the pull request:
| # | 📍 NOTE | What |  |
| - | ------- | - | - |
| 1️⃣ | This one first | Modernize gem structure | <=== You are here |
| 2️⃣ | Then see #37 next | Upgrade to v1.7 spec |  |
| 3️⃣ | Then see #39  | Add --validate feature |  |
| 4️⃣ | Then see #40  | Add --include-metadata feature |  |

🎨 modernize for 2025

### Added

- `CONTRIBUTING.md` file to help people find their way to contributing
- `CHANGELOG.md` file to document notable changes in keep-a-changelog format
- `Cyclonedx::BomHelpers` module to house helper methods, replacing global methods
- `Cyclonedx::BomBuilder` class, replacing `Bombuilder` (note the capitalization change)
- `Cyclonedx::BomComponent` class, replacing `BomComponent`
- `Cyclonedx::Ruby::Version::VERSION` constant to hold the version number (also available as `Cyclonedx::VERSION`)
- `Cyclonedx::Ruby::Deprecation` module to help manage deprecations
- dev dependency: `stone_checksums`
  - For SHA-256 and SHA-512 checksum generation for each release.
- ~~signed gem releases~~ (postponed until later)
  - See: [RubyGems Security Guide][🔒️rubygems-security-guide]
- CI matrix testing on Ruby 3.3, 3.4

[🔒️rubygems-security-guide]: https://guides.rubygems.org/security/#building-gems

### Changed

- Updated gemspec metadata for clarity and consistency
- Modernized Rakefile, dotfiles, and test setup
- `LICENSE` => `LICENSE.txt` to simplify parsing consistency on various platforms and tools
- `cucumber` v8 => v10
- `aruba` v2.1 => v2.2

### Deprecated

- `BomComponent` => `Cyclonedx::BomComponent`
- `Bombuilder` => `Cyclonedx::BomBuilder` (note the capitalization change)
- `Object.purl` => `Cyclonedx::BomHelpers.purl`
- `Object.random_urn_uuid` => `Cyclonedx::BomHelpers.random_urn_uuid`
- `Object.build_bom` => `Cyclonedx::BomHelpers.build_bom`
- `Object.build_json_bom` => `Cyclonedx::BomHelpers.build_json_bom`
- `Object.build_bom_xml` => `Cyclonedx::BomHelpers.build_bom_xml`
- `Object.get_gem` => `Cyclonedx::BomHelpers.get_gem`

### Fixed

- `Nokogiri::XML::Builder` context relies on `method_missing`
  - Globally defined `Object#purl` conflicted with `<purl>`.
  - Moved to `Cyclonedx::BomHelpers.purl` to avoid conflict in v2.0.0 (along with all other global methods)
  - Fixed existing usage via the built-in Nokogiri workaround of adding an underscore `purl_`
  - The XML tag is unchanged as `<purl>`

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🔨 binstubs: console, setup, rake, rspec, cucumber

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🚚 rename LICENSE => LICENSE.txt

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🎨 modernize dotfiles

- .gitignore
- .rspec

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🎨 modernize Rakefile

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🎨 modernize for 2025

### Added

- `CHANGELOG.md` file to document notable changes in keep-a-changelog format
- `Cyclonedx::BomHelpers` module to house helper methods, replacing global methods
- `Cyclonedx::BomBuilder` class, replacing `Bombuilder` (note the capitalization change)
- `Cyclonedx::BomComponent` class, replacing `BomComponent`
- `Cyclonedx::Ruby::Version::VERSION` constant to hold the version number (also available as `Cyclonedx::VERSION`)
- `Cyclonedx::Ruby::Deprecation` module to help manage deprecations

### Changed

- Updated gemspec metadata for clarity and consistency
- Modernized Rakefile, dotfiles, and test setup
- `LICENSE` => `LICENSE.txt` to simplify parsing consistency on various platforms and tools
- `cucumber` v8 => v10
- `aruba` v2.1 => v2.2

### Deprecated

- `BomComponent` => `Cyclonedx::BomComponent`
- `Bombuilder` => `Cyclonedx::BomBuilder` (note the capitalization change)
- `Object.purl` => `Cyclonedx::BomHelpers.purl`
- `Object.random_urn_uuid` => `Cyclonedx::BomHelpers.random_urn_uuid`
- `Object.build_bom` => `Cyclonedx::BomHelpers.build_bom`
- `Object.build_json_bom` => `Cyclonedx::BomHelpers.build_json_bom`
- `Object.build_bom_xml` => `Cyclonedx::BomHelpers.build_bom_xml`
- `Object.get_gem` => `Cyclonedx::BomHelpers.get_gem`

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🎨 modernize for 2025

### Added

- `CHANGELOG.md` file to document notable changes in keep-a-changelog format
- `Cyclonedx::BomHelpers` module to house helper methods, replacing global methods
- `Cyclonedx::BomBuilder` class, replacing `Bombuilder` (note the capitalization change)
- `Cyclonedx::BomComponent` class, replacing `BomComponent`
- `Cyclonedx::Ruby::Version::VERSION` constant to hold the version number (also available as `Cyclonedx::VERSION`)
- `Cyclonedx::Ruby::Deprecation` module to help manage deprecations

### Changed

- Updated gemspec metadata for clarity and consistency
- Modernized Rakefile, dotfiles, and test setup
- `LICENSE` => `LICENSE.txt` to simplify parsing consistency on various platforms and tools
- `cucumber` v8 => v10
- `aruba` v2.1 => v2.2

### Deprecated

- `BomComponent` => `Cyclonedx::BomComponent`
- `Bombuilder` => `Cyclonedx::BomBuilder` (note the capitalization change)
- `Object.purl` => `Cyclonedx::BomHelpers.purl`
- `Object.random_urn_uuid` => `Cyclonedx::BomHelpers.random_urn_uuid`
- `Object.build_bom` => `Cyclonedx::BomHelpers.build_bom`
- `Object.build_json_bom` => `Cyclonedx::BomHelpers.build_json_bom`
- `Object.build_bom_xml` => `Cyclonedx::BomHelpers.build_bom_xml`
- `Object.get_gem` => `Cyclonedx::BomHelpers.get_gem`

### Fixed

- `Nokogiri::XML::Builder` context relies on `method_missing`
  - Globally defined `Object#purl` conflicted with `<purl>`.
  - Moved to `Cyclonedx::BomHelpers.purl` to avoid conflict in v2.0.0 (along with all other global methods)
  - Fixed existing usage via the built-in Nokogiri workaround of adding an underscore `purl_`
  - The XML tag is unchanged as `<purl>`

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
⚡️ Frozen Strings (performance benefit)

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
⏪️ Revert accidental changes

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling commented on bin/setup:
Oh, and regarding the file, `bin/setup` and `bin/console` are now a standard part of all gems generated with `bundle gen <gem name>`.
So this PR is an attempt to bring the gem up to the modern expectations of ruby gems, generally, with a particular focus on DX.

jkowalleck replied:
please add CONTRIBUTING.md to the project.

this should include the following
- how to setup
- how to run tests
- any code styles or similar rules
- requirement to sign-off commits - similar to <https://github.com/CycloneDX/cyclonedx-python-lib/blob/5d42b558e655abd69b78a0bb1e60996c72fc6a18/CONTRIBUTING.md?plain=1#L64>
  > Please sign off your commits, to show that you agree to publish your changes under the current terms and licenses of the project , and to indicate agreement with [Developer Certificate of Origin (DCO)](https://developercertificate.org/).
  > ```shell
  > git commit --signed-off ...
  > ```

mperham commented:
I’d like to see this work integrated too. Thanks @pboling!

one request: remove unnecessary 3rd party dependencies. No real reason this lib needs nokogiri or rest-client.

pboling committed:
🚨 Linting

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🔒️ Signed releases, checksums & CONTRIBUTING.md

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
👷 Update build matrix (light)

- Bigger build matrix changes are needed to continue running against older bunlders.
- will introduce the appraisal pattern later

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🔨 binstub: gem_checksums

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

jkowalleck commented on .github/workflows/ruby.yml:
```diff
@@ -18,10 +18,14 @@ jobs:
+          - '3.4.7'
+          - '3.3.9'
+          - '3.2.9'
+# Not supported by latest bundler. These will be added back to the build matrix later.
```
note to self: create a followup ticket for this

jkowalleck replied:
ticket: https://github.com/CycloneDX/cyclonedx-ruby-gem/issues/43

jkowalleck commented on cyclonedx-ruby.gemspec:
```diff
@@ -1,26 +1,79 @@
+  spec.metadata["source_code_uri"] = "#{spec.homepage}/tree/v#{spec.version}"
+  spec.metadata["changelog_uri"] = "#{spec.homepage}/blob/v#{spec.version}/CHANGELOG.md"
+  spec.metadata["bug_tracker_uri"] = "#{spec.homepage}/issues"
+  spec.metadata["documentation_uri"] = "https://www.rubydoc.info/gems/#{spec.name}/#{spec.version}"
```
please remove, as long as we dong host any docs there.

pboling replied:
They are actually already hosted there, but the documentation isn't great in the current release. The next release will have excellent YARD documentation.

https://rubydoc.info/gems/cyclonedx-ruby

jkowalleck committed:
Merge branch 'master' into fix/nokogiri-incompatible-with-global-purl

jkowalleck committed:
add `NOTICE` to `extra_rdoc_files`

Signed-off-by: Jan Kowalleck <jan.kowalleck@gmail.com>

pboling committed:
📝 Use hardcoded homepage roots

- cleanup spec.files

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling commented on cyclonedx-ruby.gemspec:
```diff
@@ -1,26 +1,79 @@
-  spec.executables = spec.files.grep(%r{\Aexe/}) { |f| File.basename(f) }
+
+  spec.metadata["homepage_uri"] = "https://#{spec.name.tr('_', '-')}.galtzo.com/"
+  spec.metadata["source_code_uri"] = "#{spec.homepage}/tree/v#{spec.version}"
```
@jkowalleck Fixed.

pboling replied:
Fixed.

pboling commented on bom.xml:
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<bom xmlns="http://cyclonedx.org/schema/bom/1.1" version="1" serialNumber="urn:uuid:ffc51349-2d7d-408e-b2c1-3e3f220e6d2f">
+  <components>
```
Good point. It's a fixture for validation. Perhaps I'll move it into spec/fixtures...

pboling commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,151 @@
+
+Releasing and signing
+- SKIP_GEM_SIGNING: If set, skip gem signing during build/release
+- GEM_CERT_USER: Username for selecting your public cert in `certs/<USER>.pem` (defaults to $USER)
```
I can reverse it, but I do have a slight preference for signing-by-default, as part of a secure-by-default posture.

pboling commented on certs/pboling.pem:
Additionally, until there is a proper signing key host we rely on as a community, this is the next best thing. Signing gems is pointless if the signing can't be verified, and that can only be done against the public cert. I've made mine available in a special ["certs" project](https://raw.github.com/galtzo-floss/certs/main/pboling.pem) for longevity, but in order to facilitate other signers/releasers it makes sense to have a copy in the repo itself. I am certain that this pattern is in need of improvement, but I'm not sure if it is something that should be solved here / now.

pboling replied:
FWIW: This is what I envision adding to the readme at some point (made less invasive by placing in `details` tag as I do on other projects perhaps).

### 🔒 Secure Installation

For Medium or High Security Installations

This gem is cryptographically signed, and has verifiable SHA-256 and SHA-512 checksums. Be sure the gem you install hasn’t been tampered with by following the instructions below.

Add signer's public key (if you haven’t already, expires 2045-04-29) as a trusted certificate:

```console
gem cert --add <(curl -Ls https://raw.github.com/galtzo-floss/certs/main/pboling.pem)
```

You only need to do that once.  Then proceed to install with:

```console
gem install cyclonedx-ruby -P HighSecurity
```

The `HighSecurity` trust profile will verify signed gems, and not allow the installation of unsigned dependencies.

If you want to up your security game full-time:

```console
bundle config set --global trust-policy MediumSecurity
```

`MediumSecurity` instead of `HighSecurity` is necessary if not all the gems you use are signed.

NOTE: Be prepared to track down certs for signed gems and add them the same way you added this signer's.

pboling replied:
That `gem cert add` command is the core of the question here.
I have a centralized certs project so I can use that same line in every gem I publish.
This one could alternatively use:
```
gem cert --add <(curl -Ls https://raw.github.com/CycloneDX/cyclonedx-ruby-gem/main/certs/pboling.pem)
```
but only if we keep the certs in this repo.

I'm open to other ideas. The gem signing infrastructure for Ruby is admittedly lackluster. I am trying to improve it, but that's a separate effort.

pboling replied:
@jkowalleck it is worth noting that other signed gems that are developed by a team use a team cert, shared among those who publish the gem, and rotated when that set changes.  It's a small improvement, but the entire process is still very under-developed.

pboling replied:
@jkowalleck gem signing / cert logic is removed.

pboling commented on cyclonedx-ruby.gemspec:
```diff
@@ -1,26 +1,79 @@
+  # Ref: https://gitlab.com/ruby-oauth/version_gem/-/issues/3
+  # Hence, only enable signing if `SKIP_GEM_SIGNING` is not set in ENV.
+  # See CONTRIBUTING.md
+  unless ENV.include?("SKIP_GEM_SIGNING")
```
See my comments here for [more context](https://github.com/CycloneDX/cyclonedx-ruby-gem/pull/38#discussion_r2479478826) on gem signing.

pboling committed:
🔥 Remove unnecessary test artifacts

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🔥 Remove gem release signing logic

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🔥 Remove documentation uri

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

jkowalleck commented on spec/spec_helper.rb:
```diff
@@ -1,7 +1,16 @@
-# Copied from https://github.com/cucumber/aruba/blob/3b1a6cea6e3ba55370c3396eef0a955aeb40f287/spec/spec_helper.rb
-# Licensed under MIT - https://github.com/cucumber/aruba/blob/3b1a6cea6e3ba55370c3396eef0a955aeb40f287/LICENSE
```
why was this license note removed? 
the license text clearly expresses, that the license information MUST NOT removed, right?

pboling replied:
A license header like this is not applicable to a file name, but to the contents of the file. It is very clear that the code was copied from a specific place (Aruba) and inherited the license as a result of that.

Removing a license header like this is simply a matter of deleting all code in the file and writing it from scratch, which I did, and which is needed, because this project's spec helper will have nothing in common with Aruba's spec_helper.

Thus the contents of the file revert to being under the same license as the core project.

I'd be happy to add back a license header, but I do note that license headers are not universally applied in the project. Of course I'd need to add it back without the reference to aruba, since it no longer has any relevance.

pboling committed:
🐛 Fix latent reference to BomComponent without namespace

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
🐛 Fix latent reference to @logger out of context

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
📄 Add back license, scoped to relevant lines

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling commented on spec/spec_helper.rb:
```diff
@@ -11,6 +20,8 @@
-Dir.glob(File.expand_path('support/**/*.rb', __dir__)).sort.each { |f| require_relative f }
+mimic_next_major = ENV.fetch('MIMIC_NEXT_MAJOR_VERSION', 'false')
+# Require via legacy path until v2.0.0, and unless testing functionality in preparation for next major release
+require 'bom_builder' if mimic_next_major.casecmp?('false')
```
Not going to change it.  `Unless` has higher cognitive load than `if` generally, and the comment is a combination logic clause which is not entirely represented by the code logic below, so it's not a 1:1 relationship to start with.

pboling committed:
🐛 fix reference to variable bom_directory within rescue block

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

pboling committed:
✏️ fix method indicator in deprecation warnings

Signed-off-by: Peter H. Boling <peter.boling@gmail.com>

jkowalleck commented on Gemfile.lock:
```diff
@@ -0,0 +1,196 @@
+  remote: .
+  specs:
+    cyclonedx-ruby (1.2.0)
+      activesupport (~> 7.0)
```
ticket : https://github.com/CycloneDX/cyclonedx-ruby-gem/issues/42

jkowalleck commented:
thank you so much for this, @pboling .

thank you for reviewing, @seuros .
